src/shared/shared_ck/particle_dynamics/fluid_dynamics/boundary_condition/fluid_boundary_ck.h

```diff
@@ -254,6 +256,76 @@ class TagBufferParticlesCK : public BaseLocalDynamics<AlignedBoxPartByCell>
     DiscreteVariable<Vecd> *dv_pos_;
     DiscreteVariable<int> *dv_buffer_particle_indicator_;
 };
+
+class FlowrateCalculateCK
+    : public BaseLocalDynamicsReduce<ReduceSum<Vec3d>, AlignedBoxPartByCell>
+{
+  public:
+    // Constructor: note that the reduced quantity is of type Vec3d.
+    FlowrateCalculateCK(AlignedBoxPartByCell &aligned_box_part)
+        : BaseLocalDynamicsReduce<ReduceSum<Vec3d>, AlignedBoxPartByCell>(aligned_box_part),
+          sv_aligned_box_(aligned_box_part.svAlignedBox()),
+          dv_pos_(particles_->template getVariableByName<Vec3d>("Position")),
+          dv_vel_(particles_->template getVariableByName<Vec3d>("Velocity"))
+    {
+    }
+
+    virtual ~FlowrateCalculateCK() {}
+
+    // The ReduceKernel: for each particle, return its velocity if inside the box; else, zero.
+    class ReduceKernel
+    {
+      public:
+        template <class ExecutionPolicy, class EncloserType>
+        ReduceKernel(const ExecutionPolicy &ex_policy, EncloserType &encloser)
+            : aligned_box_(encloser.sv_aligned_box_->DelegatedData(ex_policy)),
+              pos_(encloser.dv_pos_->DelegatedData(ex_policy)),
+              vel_(encloser.dv_vel_->DelegatedData(ex_policy))
+        {
+        }
+
+        Vec3d reduce(size_t index_i, Real dt = 0.0)
+        {
+            Vec3d result = Vec3d::Zero();
+            if (aligned_box_->checkUpperBound(pos_[index_i]) &&
```

Comment:
here, you say, if the particle is not within aligned box. It should use checkInBound()?